### DISCUSSION TITLE
🚨 v3.9.x

### DISCUSSION THREAD

**Discussion by trexfeathers** [SciTools/iris#5850] [Releases]
An impromptu release to get two critical features out as quickly as possible:

- #5833
- #5740 

| Manager | Buddy | Release Branch | Project |
| :---: | :---: | :---: | :---: |
| @trexfeathers | ? | [v3.9.x](https://github.com/SciTools/iris/tree/v3.9.x) | - |

| Minor Release | Status | Date | RTD | Release | conda-forge | PyPI |
| :---: | :---: | :---: | :---: | :---: | :---: | :---: |
| `v3.9.0rc0` | [![status](https://img.shields.io/badge/v3.9.0rc0-released-green)](https://github.com/SciTools/iris/discussions/5850) | `03-04-2024` | [Docs](https://scitools-iris.readthedocs.io/en/v3.9.0rc0/index.html) | [v3.9.0rc0](https://github.com/SciTools/iris/releases/tag/v3.9.0rc0) | [rc_iris](https://anaconda.org/conda-forge/iris?version=3.9.0rc0&channel=rc_iris) | [scitools-iris (--pre)](https://pypi.org/project/scitools-iris/3.9.0rc0/) |
| `v3.9.0` | [![status](https://img.shields.io/badge/v3.9.0-released-green)](https://github.com/SciTools/iris/discussions/5850) | `22-04-2024` | [Docs](https://scitools-iris.readthedocs.io/en/v3.9.0/index.html) | [v3.9.0](https://github.com/SciTools/iris/releases/tag/v3.9.0) | [main](https://anaconda.org/conda-forge/iris?version=3.9.0&channel=main) | [scitools-iris](https://pypi.org/project/scitools-iris/3.9.0/) |

**Comment by trexfeathers** [comment]
In our experience, any significant change to Iris' use of Dask has some unexpected consequences, as well as expected ones!. These are normally only noticeable to people doing big operations with lots of memory and many CPUs. So we're looking for volunteers who might be able to apply `v3.9.0rc0` to their existing workflows to make sure nothing bad happens.

**Reply by trexfeathers** [reply]
@ESMValGroup/esmvaltool-developmentteam any interest?

**Reply by trexfeathers** [reply]
@matthew-mizielinski, @Jon-Lillis I'm told you might have some candidate workflows?

**Reply by Jon-Lillis** [reply]
Yep - I've got a standalone jupyter notebook that demonstrates the problem of failing to to save a large cube. Could be related to this issue, or could just be me being silly. Happy to share it if it helps.

**Reply by matthew-mizielinski** [reply]
@piotr-florek-mohc, fancy building a CDDS environment with an update to iris, which would be good to do for CDDS v3.0 anyway, and rerunning some of the N512 tests in the SPICE testing?

@trexfeathers, note that our tests will do a fair amount of IO over an extended job (6-12 hours), but is only a single threaded application. We use iris to read data, but the writing out is done via the PCMDI/cmor library.

**Comment by trexfeathers** [comment]
📢 Release candidate is available NOW

**Reply by piotr-florek-mohc** [reply]
I ran one of our benchmarks we used for comparative testing of SPICE and Azure, and it OOMed during processing of one of larger, 3D diagnostics (`cl`).

```
Command terminated by signal 9
	Command being timed: "mip_convert mip_convert.cfg.2024-04-09T1721Z -a -s ap6 -l mip_convert --datestamp 2024-04-09T1721Z --mip_era CMIP6"
	User time (seconds): 179.49
	System time (seconds): 331.44
	Percent of CPU this job got: 17%
	Elapsed (wall clock) time (h:mm:ss or m:ss): 48:29.64
	Average shared text size (kbytes): 0
	Average unshared data size (kbytes): 0
	Average stack size (kbytes): 0
	Average total size (kbytes): 0
	Maximum resident set size (kbytes): 24798960
	Average resident set size (kbytes): 0
	Major (requiring I/O) page faults: 3892
	Minor (reclaiming a frame) page faults: 1626138
	Voluntary context switches: 508683
	Involuntary context switches: 136769
	Swaps: 0
	File system inputs: 533102424
	File system outputs: 766432
	Socket messages sent: 0
	Socket messages received: 0
	Signals delivered: 0
	Page size (bytes): 4096
	Exit status: 0
```

Since I had to update quite a few dependencies to install the rc, I will try to replicate the environment with older iris and check if it works better.

**Reply by trexfeathers** [reply]
@piotr-florek-mohc thanks

**Comment by piotr-florek-mohc** [comment]
corresponding output from successful job using iris 3.4.1 I ran back in March:

```
command being timed: "mip_convert mip_convert.cfg.2024-03-22T075832Z_723 -a -s ap6 -l /data/scratch/piotr.florek/azure_test_case/mip_convert_logs/mip_convert_log_723 --datestamp 2024-03-22T075832Z_723 --mip_era CMIP6"
	User time (seconds): 6430.68
	System time (seconds): 1122.13
	Percent of CPU this job got: 65%
	Elapsed (wall clock) time (h:mm:ss or m:ss): 3:13:30
	Average shared text size (kbytes): 0
	Average unshared data size (kbytes): 0
	Average stack size (kbytes): 0
	Average total size (kbytes): 0
	Maximum resident set size (kbytes): 22091400
	Average resident set size (kbytes): 0
	Major (requiring I/O) page faults: 456230
	Minor (reclaiming a frame) page faults: 48120811
	Voluntary context switches: 1637597
	Involuntary context switches: 256478
	Swaps: 0
	File system inputs: 1026081408
	File system outputs: 196527200
	Socket messages sent: 0
	Socket messages received: 0
	Signals delivered: 0
	Page size (bytes): 4096
	Exit status: 0
```

All jobs were run with `--mem=25G`

**Reply by pp-mo** [reply]
Well it looks like that is using ~22/25Gb, obvioulsy close to the edge !

Can we easily try different things here?
E.G. it would be good to (say) double the memory allocation + retry the latest rc to see what it now costs.
TBH I'm not sure why it should use that much memory anyway -- we maybe need a close look at what your code is doing.

At Iris 3.6 we overhauled netcdf saving, which adds some complexity but supports[ use with distributed, and 'lazy deferred saving'](https://scitools-iris.readthedocs.io/en/stable/whatsnew/3.6.html#v3-6-18-may-2023).  
_However_, that also introduced **a serious memory bug**, which we only recently became aware of, and is finally [fixed in v3.9](https://github.com/SciTools/iris/issues/5753).  Basically, in various cases saving a cube may load the whole of its data to memory (which should not happen).

So if your 'c1' is bigger than memory, it would probably blow up with any of Iris 3.6-3.8, anyway, but the latest should "fix" that.  Meanwhile though, it's entirely likely that Iris 3.9 would be **somewhat** more memory hungry than 3.4, since the revamped approach is overall that much more complex, and I think the way that Dask works may be generally less memory-efficient with the later code.

All the above _assuming_ that saving is responsible for the big memory cost.  Of course you might be realising some big cubes anyway?

**Reply by piotr-florek-mohc** [reply]
Hi Patrick,
yes I'm rerunning the job with 30GB limit, so we should soon know how large iris' overhead is. In terms of data volume the cube is 85x768x1024 but I think our code processes 6 months worth of data (i.e. 180 daily time points) in one go; will double check that.

**Reply by piotr-florek-mohc** [reply]
fail, repeating with 40G now

**Reply by matthew-mizielinski** [reply]
~~Piotr will post some of the details later, but initial results suggest we need to increase the amount of memory allocated for this test job by a factor of 7 in order to get our test case to work. The maximum resident set size has gone from 22 GB shown above, to 154 GB when changing from iris 3.4.1 to the 3.9.0rc*.  We're re-running a number of reference jobs now to confirm that it isn't something on the system that has unexpectedly changed to give rise to this, and also looking into some of the darker corners of our code base to understand what is going on.~~

~~Note that the MIP Convert tool being used for this test doesn't make any direct use of dask and we do not use multiple threads. While data is read in by iris, it is written out via the CMOR library to enforce CMIP standards compliance so we are not using any of the iris netcdf saver code.~~

~~*In stepping up to the newer iris version we've had to update the python version to 3.11 and it is not impossible that other dependencies changing in our conda environment could be involved in this expansion of memory use.  We were planning on picking up iris v3.7.1 in the near future, and we'll hopefully run some similar test cases in the next week or two.~~

Just in case anyone comes across this again; there was an error in our testing that resulted in loading 6 times more data than in our reference case which was _not_ due to this change to iris (see further details from Piotr below).

**Reply by trexfeathers** [reply]
~~Can I clarify if your scripts are using Iris saving at any point? That's the only thing we changed in 3.9~~

Sorry for skim-reading, you clearly already told us that you are NOT.

**Reply by piotr-florek-mohc** [reply]
Good morning, Matt reminded me this morning that there might have been a code change in our codebase (subannual slicing) that I was sure hadn't been deployed, but on double checking I found out I did actually partially implement it for the purpose of SPICE/Azure benchmarks (long story short: I couldn't get it working for Gregorian calendar, but since tests were to be done using our standard model configuration, I could get away with 360 calendar). I forgot to patch it when doing the 3.9.x installation, and that explains why the code OOMs now.

I merged the change and I restarted the benchmark, fingers crossed it will behave more reasonably now with loading 6 times less data into memory. Apologies for all the confusion.

**Reply by piotr-florek-mohc** [reply]
I've run our tests with a fixed workflow and can confirm that the silly behaviour encountered before does not occur. The maxRSS is still a bit higher than before (28G vs 22G) but not egregiously so. Once again, apologies for incurring confusion!

**Comment by trexfeathers** [comment]
📢 `v3.9.0` is available NOW
